src/eloelo/elodisco/hero_pool_generators.rs

```
@@ -0,0 +1,668 @@
+        vec![
+            HeroTag("Core".to_string()),
+            HeroTag("Support".to_string()),
+            HeroTag("Carry".to_string()),
+        ][tag_id % 3]
+            .clone()
+    }
+}
+
+impl TryFrom<String> for HeroTag {
+    type Error = Error;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        if HeroTag::all().contains(value.as_str()) {
+            Ok(HeroTag(value))
+        } else {
+            Err(format_err!(
+                "Incorrect hero name: \"{}\". See `all` for list of valid names.",
+                value
+            ))
+        }
+    }
+}
+
+#[cfg(test)]
+impl From<&str> for HeroTag {
+    fn from(value: &str) -> Self {
+        HeroTag::try_from(String::from(value)).unwrap()
+    }
+}
+
+impl Display for HeroTag {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl Borrow<str> for HeroTag {
+    fn borrow(&self) -> &str {
+        &self.0
+    }
+}
+
+impl AsRef<HeroTag> for HeroTag {
+    fn as_ref(&self) -> &HeroTag {
+        &self
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct TaggedHeroPool {
+    pub tags: HashMap<HeroTag, HashSet<Hero>>,
+    pub hero_assignement: HashMap<PlayerInfo, Vec<Hero>>,
+    pub taken: HashSet<Hero>,
+}
+
+impl TaggedHeroPool {
+    pub fn new() -> Self {
+        TaggedHeroPool {
+            tags: Self::read_tags(),
+            ..Default::default()
+        }
+    }
+
+    fn read_tags() -> HashMap<HeroTag, HashSet<Hero>> {
+        include_str!("dota_heroes.csv")
+            .split("\n")
+            .map(|s| {
+                let raw: Vec<&str> = s.split(",").map(|s| s.trim()).collect();
+                (raw[0].to_string(), raw[1].to_string())
+            })
+            .fold(HashMap::new(), |mut m, (hero_name, tag_name)| {
+                if let Some(hero) = Hero::try_from(hero_name).ok() {
+                    if let Some(tag) = HeroTag::try_from(tag_name).ok() {
```

It's better to panic if Hero or Hero tag cannot be created, since the parsed data is baked into the binary anyway.

Make sure we have a test exercising this code so it fails early.

---

Every UT calls this. Done